Mathlib/MeasureTheory/Measure/Haar/Unique.lean

```diff
@@ -282,56 +302,466 @@ lemma integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport
     exact (exists_integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ' μ).choose_spec
       f hf h'f
 
-/-- The scalar factor between two left-invariant measures is non-zero when both measures are
-positive on open sets. -/
 @[to_additive]
-lemma haarScalarFactor_pos_of_isOpenPosMeasure (μ' μ : Measure G) [IsFiniteMeasureOnCompacts μ]
-    [IsFiniteMeasureOnCompacts μ'] [IsMulLeftInvariant μ] [IsMulLeftInvariant μ']
-    [IsOpenPosMeasure μ] [IsOpenPosMeasure μ'] : 0 < haarScalarFactor μ' μ := by
-  rw [pos_iff_ne_zero]
-  intro H
+lemma haarScalarFactor_eq_mul (μ' μ ν : Measure G)
+    [IsHaarMeasure μ] [IsHaarMeasure ν] [IsFiniteMeasureOnCompacts μ'] [IsMulLeftInvariant μ'] :
+    haarScalarFactor μ' ν = haarScalarFactor μ' μ * haarScalarFactor μ ν := by
   -- The group has to be locally compact, otherwise the scalar factor is 1 by definition.
   by_cases hG : LocallyCompactSpace G; swap
-  · simp [haarScalarFactor, hG] at H
+  · simp [haarScalarFactor, hG]
   -- Fix some nonzero continuous function with compact support `g`.
-  obtain ⟨g, g_cont, g_comp, g_nonneg, g_one⟩ :
-      ∃ (g : G → ℝ), Continuous g ∧ HasCompactSupport g ∧ 0 ≤ g ∧ g 1 ≠ 0 := by
-    rcases exists_compact_mem_nhds (1 : G) with ⟨k, hk, k_mem⟩
-    rcases exists_continuous_one_zero_of_isCompact hk isClosed_empty (disjoint_empty k)
-      with ⟨⟨g, g_cont⟩, gk, -, g_comp, hg⟩
-    refine ⟨g, g_cont, g_comp, fun x ↦ (hg x).1, ?_⟩
-    have := gk (mem_of_mem_nhds k_mem)
-    simp only [ContinuousMap.coe_mk, Pi.one_apply] at this
-    simp [this]
-  have int_g_pos : 0 < ∫ x, g x ∂μ' := by
+  obtain ⟨⟨g, g_cont⟩, g_comp, g_nonneg, g_one⟩ :
+    ∃ (g : C(G, ℝ)), HasCompactSupport g ∧ 0 ≤ g ∧ g 1 ≠ 0 := exists_continuous_nonneg_pos 1
+  have Z := integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ' μ g_cont g_comp
+  simp only [integral_smul_nnreal_measure, smul_smul,
+    integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ' ν g_cont g_comp,
+    integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ ν g_cont g_comp] at Z
+  have int_g_pos : 0 < ∫ x, g x ∂ν := by
     apply (integral_pos_iff_support_of_nonneg g_nonneg _).2
-    · exact IsOpen.measure_pos μ' g_cont.isOpen_support ⟨1, g_one⟩
+    · exact IsOpen.measure_pos ν g_cont.isOpen_support ⟨1, g_one⟩
     · exact g_cont.integrable_of_hasCompactSupport g_comp
-  have := integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport μ' μ g_cont g_comp
-  simp only [H, zero_smul, integral_zero_measure] at this
-  linarith
+  change (haarScalarFactor μ' ν : ℝ) * ∫ (x : G), g x ∂ν =
+    (haarScalarFactor μ' μ * haarScalarFactor μ ν : ℝ≥0) * ∫ (x : G), g x ∂ν at Z
+  simpa only [mul_eq_mul_right_iff (M₀ := ℝ), int_g_pos.ne', or_false, NNReal.eq_iff] using Z
+
```

Comment:
Is it possible to add the fact that haarScalarFactor multiplies when you have 3 measures? (Maybe not, because of the two different hypotheses, IsHaarMeasure and IsLeftInvariantMeasure.)